src/mpyl/stages/discovery.py

```diff
@@ -60,159 +77,139 @@ def __has_invalidated_dependencies(
+                f"is touched"
             )
-            __log_invalidation(logger, project, stage, path, message)
             return True
 
     return False
 
 
-def is_invalidated(
-    logger: logging.Logger,
-    project: Project,
-    stage: str,
-    changed_file: ChangedFile,
-    change_history: list[Revision],
-    steps: StepsCollection,
-) -> bool:
-    invalidated_by_file = is_invalidated_by_file(
-        logger, project, stage, changed_file, steps
-    )
-    if invalidated_by_file:
-        output = Output.try_read(project.target_path, stage)
-        return stage == deploy.STAGE_NAME or _is_output_invalid(
-            logger, output, change_history, changed_file.revision
-        )
-    return False
-
-
-def _is_output_invalid(
-    logger: logging.Logger,
-    output: Optional[Output],
-    change_history: list[Revision],
-    changed_file_revision: str,
-) -> bool:
-    if output is None:
-        return True
-    if not output.success:
-        return True
-
-    artifact = output.produced_artifact
-    if artifact is None:
-        return True
-
-    is_newer = _revision_is_newer_than_artifact(
-        artifact.revision, changed_file_revision, change_history
-    )
-
-    valid = "❌ stale" if is_newer else "✅ valid"
-    newer = "newer" if is_newer else "older"
-
-    message = (
-        f"'{artifact.artifact_type.name}' produced by '{artifact.producing_step}' is {valid} "
-        f"because {changed_file_revision} is {newer} than {artifact.revision}  "
-    )
-    logger.debug(message)
-
-    return is_newer
-
-
-def is_invalidated_by_file(
-    logger: logging.Logger,
-    project: Project,
-    stage: str,
-    changed_file: ChangedFile,
-    steps: Optional[StepsCollection],
-) -> bool:
-    if changed_file.path.startswith(project.root_path):
-        message = f"it is under the project root '{project.root_path}'"
-        __log_invalidation(logger, project, stage, changed_file.path, message)
-        return True
-
-    deps: Optional[Dependencies] = project.dependencies
-    if deps is None:
+def is_output_cached(output: Optional[Output], cache_key: str) -> bool:
+    if (
+        output is None
+        or not output.success
+        or output.produced_artifact is None
+        or not output.produced_artifact.hash
+    ):
         return False
-    return __has_invalidated_dependencies(
-        logger, deps, project, stage, changed_file.path, steps
-    )
+    return output.produced_artifact.hash == cache_key
 
 
-def _revision_is_newer_than_artifact(
-    artifact_hash: str, file_hash: str, change_history: list[Revision]
-):
-    hashes = {rev.hash for rev in change_history}
-    if artifact_hash not in hashes:
-        return True
-
-    for rev in sorted(change_history, key=lambda c: c.ord):
-        if rev.hash == file_hash:
-            return False
-        if rev.hash == artifact_hash:
-            return True
-    return True
-
+def hashed_changes(files: set[str]) -> str:
```

Comment:
It's a bit weird to me that the hashing happens when we figure out the build plan. Shouldn't we hash when we're done with a stage? Maybe I'm not getting the new approach fully yet :D

Reply:
We need to hash again to be able to compare to the stored hash in `is_output_cached`. But maybe @ptab can explain this better.

Reply:
Every time we trigger a build with mpyl we hash the content of the files modified in the branch or since a previous tag, so we know whether we need to fully re-run the stage or use the results from a previous run.

The set of modified files does not change throughout a run, so it really doesn't matter when we calculate the hash as long as it's before the check @Jorg88 mentioned. We could even take this hashing completely outside MPyL and pass it as a run parameter (@Jorg88 and I are considering this as a future improvement).

Reply:
got it. just for readability I would move it to a separate helper module and write specific tests for it (like the hash changes if a new file is added, the hash goes back to what it was before when that new file is not changed anymore, stuff like that)

Reply:
Good suggestions 👍